src/VisualStudio/Core/Def/ProjectSystem/MetadataReferences/VisualStudioMetadataReferenceManager.cs

```diff
@@ -246,75 +247,77 @@ private bool TryCreateAssemblyMetadataFromMetadataImporter(FileKey fileKey, [Not
             {
-                // here, we don't care about timestamp since all those bits should be part of Fx. and we assume that 
-                // it won't be changed in the middle of VS running.
-                var fullPath = fileKey.FullPath;
+                // We might not be able to use COM services to get this if VS is shutting down. We'll synchronize to make sure this
+                // doesn't race against 
+                using (_readerWriterLock.DisposableRead())
+                {
+                    // here, we don't care about timestamp since all those bits should be part of Fx. and we assume that 
+                    // it won't be changed in the middle of VS running.
+                    var fullPath = fileKey.FullPath;
 
-                info = null;
-                pImage = default;
-                length = default;
+                    info = null;
+                    pImage = default;
+                    length = default;
 
-                if (SmartOpenScopeServiceOpt == null)
-                {
-                    return false;
-                }
+                    if (SmartOpenScopeServiceOpt == null)
+                    {
+                        return false;
+                    }
 
-                if (ErrorHandler.Failed(SmartOpenScopeServiceOpt.OpenScope(fullPath, (uint)CorOpenFlags.ReadOnly, s_IID_IMetaDataImport, out var ppUnknown)))
-                {
-                    return false;
-                }
+                    if (ErrorHandler.Failed(SmartOpenScopeServiceOpt.OpenScope(fullPath, (uint)CorOpenFlags.ReadOnly, s_IID_IMetaDataImport, out var ppUnknown)))
+                    {
+                        return false;
+                    }
 
-                info = ppUnknown as IMetaDataInfo;
-                if (info == null)
-                {
-                    return false;
-                }
+                    info = ppUnknown as IMetaDataInfo;
+                    if (info == null)
+                    {
+                        return false;
+                    }
 
-                return ErrorHandler.Succeeded(info.GetFileMapping(out pImage, out length, out var mappingType)) && mappingType == CorFileMapping.Flat;
+                    return ErrorHandler.Succeeded(info.GetFileMapping(out pImage, out length, out var mappingType)) && mappingType == CorFileMapping.Flat;
+                }
             }
         }
 
         /// <exception cref="IOException"/>
         /// <exception cref="BadImageFormatException" />
         private static AssemblyMetadata CreateAssemblyMetadata(
-            FileKey fileKey, ModuleMetadata manifestModule, List<TemporaryStorageService.TemporaryStreamStorage>? storages,
-            Func<FileKey, List<TemporaryStorageService.TemporaryStreamStorage>?, ModuleMetadata> moduleMetadataFactory)
+            FileKey fileKey,
+            ModuleMetadata manifestModule,
+            Func<FileKey, ModuleMetadata> moduleMetadataFactory)
```

Comment:
i made thsi helper not care about `List<TemporaryStorageService.TemporaryStreamStorage>? storages` anymore.  Instead, the callers can encode that dependency in the lambda they pass into `moduleMetadataFactory`